bugsnag-android-core/src/main/java/com/bugsnag/android/BreadcrumbState.kt

```diff
@@ -0,0 +1,78 @@
+package com.bugsnag.android
+
+import java.io.IOException
+import java.io.StringWriter
+import java.util.Observable
+import java.util.Queue
+import java.util.concurrent.ConcurrentLinkedQueue
+
+internal class BreadcrumbState(maxBreadcrumbs: Int) : Observable(),
+    JsonStream.Streamable {
+    val store: Queue<Breadcrumb> = ConcurrentLinkedQueue()
+
+    private val maxBreadcrumbs: Int
+
+    init {
+        when {
+            maxBreadcrumbs > 0 -> this.maxBreadcrumbs = maxBreadcrumbs
+            else -> this.maxBreadcrumbs = 0
+        }
+    }
+
+    @Throws(IOException::class)
+    override fun toStream(writer: JsonStream) {
+        pruneBreadcrumbs()
+        writer.beginArray()
+        store.forEach { it.toStream(writer) }
+        writer.endArray()
+    }
+
+    fun add(breadcrumb: Breadcrumb) {
+        try {
+            if (payloadSize(breadcrumb) > MAX_PAYLOAD_SIZE) {
+                Logger.warn("Dropping breadcrumb because payload exceeds 4KB limit")
+                return
+            }
+            store.add(breadcrumb)
+            pruneBreadcrumbs()
+            setChanged()
+            notifyObservers(
+                NativeInterface.Message(
+                    NativeInterface.MessageType.ADD_BREADCRUMB, breadcrumb
+                )
+            )
+        } catch (ex: IOException) {
+            Logger.warn("Dropping breadcrumb because it could not be serialized", ex)
+        }
```

Comment:
Suggest moving this `catch` either closer to the `payloadSize` call or modifying the method to make it a validation function?

Reply:
I agree - the catch should be more specific than it was historically written as, I'll update this.